File: base_v4/BaseAdapterPresenter.java
package com.dev.rahul.basefilesamples.base_v4;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by rahul on 20/1/18.
 */

public abstract class BaseAdapterPresenter<V extends IBaseHolderView, E extends Object>
        implements IBaseAdapterPresenter<E> {

    private List<E> list;
    private BaseAdapter adapter;
    private V baseHolderView;

    public BaseAdapterPresenter() {
        this.list = new ArrayList<>();
    }

    public BaseAdapterPresenter(List<E> list) {
        this.list = list;
    }

    public void init(List<E> list) {
        this.list = list;
        if (adapter != null)
            adapter.notifyDataSetChanged();
    }

    @Override
    public void onAttach(IBaseHolderView baseHolderView) {
        this.baseHolderView = (V) baseHolderView;
    }

    public V view() {
        return baseHolderView;
    }

    @Override
    public void onAttachAdapter(BaseAdapter adapter) {
        this.adapter = adapter;
    }

    @Override
    public int getCount() {
        return list != null ? list.size() : 0;
    }

    @Override
    public E getElementAtPos(int position) {
        return list.get(position);
    }

    @Override
    public void addNewItems(List<E> listNewItems) {
        int currentSize = getCount();
        list.addAll(listNewItems);
        adapter.notifyItemRangeInserted(currentSize, getCount());
    }

    @Override
    public void addItem(E data) {
        list.add(data);
        adapter.notifyItemInserted(getCount());
    }

    @Override
    public void removeItem(int position) {
        list.remove(position);
        adapter.notifyItemRemoved(position);
    }

    @Override
    public List<E> getAllElements() {
        return list;
    }
}

File: basefiles/src/main/java/com/rp/basefiles/BaseAdapterPresenter.kt
package com.rp.basefiles

import android.support.v7.recyclerview.extensions.AsyncListDiffer
import android.support.v7.util.DiffUtil
import com.rp.util.adapter.RAdapterAsyncDiffCallback
import com.rp.util.adapter.RAdapterPayloadWatcher


abstract class BaseAdapterPresenter<V : IBaseHolderView, E> : IBaseAdapterPresenter<E> {

    private var baseHolderView: V? = null

    private var mDiffer: AsyncListDiffer<E>? = null
    private var DIFF_CALLBACK: DiffUtil.ItemCallback<E>? = null

    init {
        this.DIFF_CALLBACK = RAdapterAsyncDiffCallback<E>()
    }

    override fun onAttach(baseHolderView: IBaseHolderView) {
        this.baseHolderView = baseHolderView as V
    }

    override fun onAttachAdapter(adapter: BaseAdapter<*, *>) {
        mDiffer = AsyncListDiffer<E>(adapter, DIFF_CALLBACK!!)
    }

    override fun onAttachDiffCallback(callback: DiffUtil.ItemCallback<E>) {
        this.DIFF_CALLBACK = callback
    }

    override fun onAttachPayloadWatcher(payloadWatcher: RAdapterPayloadWatcher<E>) {
        if (DIFF_CALLBACK is RAdapterAsyncDiffCallback<*>)
            (DIFF_CALLBACK as RAdapterAsyncDiffCallback<E>)
                    .addPayloadWatcher(payloadWatcher)
    }

    fun view(): V? {
        return baseHolderView
    }

    override fun update(list: List<E>) {
        mDiffer!!.submitList(list)
    }

    override val count: Int
        get() = mDiffer?.currentList?.size ?: 0

    override fun addNewList(listNewItems: List<E>) {
        val oldList = ArrayList<E>(mDiffer!!.currentList)
        oldList.addAll(listNewItems)
        mDiffer!!.submitList(oldList)
    }

    override fun addItem(data: E) {
        val oldList = ArrayList<E>(mDiffer!!.currentList)
        oldList.add(data)
        mDiffer!!.submitList(oldList)
    }

    override fun addItemAt(position: Int, data: E) {
        val oldList = ArrayList<E>(mDiffer!!.currentList)
        oldList.add(position, data)
        mDiffer!!.submitList(oldList)
    }

    override fun removeItem(position: Int) {
        val oldList = ArrayList<E>(mDiffer!!.currentList)
        oldList.removeAt(position)
        mDiffer!!.submitList(oldList)
    }

    override fun getFrom(position: Int): E {
        return mDiffer!!.currentList.get(position)
    }

    override val all: List<E>
        get() = mDiffer?.currentList as List<E>
}